<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ scatools</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scatools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.1.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/scatools.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/TCGA.html">TGCA Sample Vignettes</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mjz1/scatools/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="EM2.html">EM2()</a></code>

        </dt>
        <dd>Iterative phasing and theta_hat estimation</dd>
      </dl><dl><dt>

          <code><a href="add_gc_cor.html">add_gc_cor()</a></code>

        </dt>
        <dd>Add GC correction to SCE Object</dd>
      </dl><dl><dt>

          <code><a href="add_gc_freq.html">add_gc_freq()</a></code>

        </dt>
        <dd>Add GC frequency</dd>
      </dl><dl><dt>

          <code><a href="add_hmmcopy.html">add_hmmcopy()</a></code>

        </dt>
        <dd>Add HMMCopy results to sce object</dd>
      </dl><dl><dt>

          <code><a href="bin_atac_frags.html">bin_atac_frags()</a></code>

        </dt>
        <dd>Bin atac fragments</dd>
      </dl><dl><dt>

          <code><a href="bin_frags_chr.html">bin_frags_chr()</a></code>

        </dt>
        <dd>Bin scATAC fragments</dd>
      </dl><dl><dt>

          <code><a href="bin_snp_data.html">bin_snp_data()</a></code>

        </dt>
        <dd>Bin SNPs into genomic segments</dd>
      </dl><dl><dt>

          <code><a href="bin_snp_data2.html">bin_snp_data2()</a></code>

        </dt>
        <dd>Bin SNP Data</dd>
      </dl><dl><dt>

          <code><a href="bind_sublist.html">bind_sublist()</a></code>

        </dt>
        <dd>Bind sublists within lists</dd>
      </dl><dl><dt>

          <code><a href="bins_10mb.html">bins_10mb</a></code>

        </dt>
        <dd>10mb Bins for HG38</dd>
      </dl><dl><dt>

          <code><a href="calc_ai.html">calc_ai()</a></code>

        </dt>
        <dd>Calculate allelic imbalance</dd>
      </dl><dl><dt>

          <code><a href="calc_allelic.html">calc_allelic()</a></code>

        </dt>
        <dd>Calculate allelic information over cells and bins</dd>
      </dl><dl><dt>

          <code><a href="calc_clonal_diversity.html">calc_clonal_diversity()</a></code>

        </dt>
        <dd>Calculate clonal diversity</dd>
      </dl><dl><dt>

          <code><a href="calc_cnv_score.html">calc_cnv_score()</a></code>

        </dt>
        <dd>Calculate CNV Score</dd>
      </dl><dl><dt>

          <code><a href="calc_snn_specificity.html">calc_snn_specificity()</a></code>

        </dt>
        <dd>Calculate SNN Specificity</dd>
      </dl><dl><dt>

          <code><a href="cluster_seurat.html">cluster_seurat()</a></code>

        </dt>
        <dd>Cluster using Seurat</dd>
      </dl><dl><dt>

          <code><a href="cnaHeatmap.html">cnaHeatmap()</a></code> <code><a href="cnaHeatmap.html">cloneCnaHeatmap()</a></code>

        </dt>
        <dd>Plot copy number heatmap</dd>
      </dl><dl><dt>

          <code><a href="correct_atac_bias.html">correct_atac_bias()</a></code>

        </dt>
        <dd>Correct ATAC Bias</dd>
      </dl><dl><dt>

          <code><a href="disjoint_bins_map.html">disjoint_bins_map()</a></code>

        </dt>
        <dd>Map between bins</dd>
      </dl><dl><dt>

          <code><a href="do_qc.html">do_qc()</a></code>

        </dt>
        <dd>Basic QC</dd>
      </dl><dl><dt>

          <code><a href="filter_sce.html">filter_sce()</a></code>

        </dt>
        <dd>Bin and/or Cell-wise quality filtering</dd>
      </dl><dl><dt>

          <code><a href="gc_cor_modal.html">gc_cor_modal()</a></code>

        </dt>
        <dd>Modal regression GC Correction for single cell data</dd>
      </dl><dl><dt>

          <code><a href="gc_modal_qc_filter.html">gc_modal_qc_filter()</a></code>

        </dt>
        <dd>GC modal QC filter</dd>
      </dl><dl><dt>

          <code><a href="get_assay_dat.html">get_assay_dat()</a></code>

        </dt>
        <dd>Utility to pull assay data and merge with features</dd>
      </dl><dl><dt>

          <code><a href="get_blacklist.html">get_blacklist()</a></code>

        </dt>
        <dd>Get blacklist regions</dd>
      </dl><dl><dt>

          <code><a href="get_chr_arm_bins.html">get_chr_arm_bins()</a></code>

        </dt>
        <dd>Get chromosome arm bins</dd>
      </dl><dl><dt>

          <code><a href="get_cytobands.html">get_cytobands()</a></code>

        </dt>
        <dd>Get genome cytobands</dd>
      </dl><dl><dt>

          <code><a href="get_f_idx.html">get_f_idx()</a></code>

        </dt>
        <dd>Get factor indices</dd>
      </dl><dl><dt>

          <code><a href="get_gene_copy.html">get_gene_copy()</a></code>

        </dt>
        <dd>Get Gene Copies</dd>
      </dl><dl><dt>

          <code><a href="get_ideal_mat.html">get_ideal_mat()</a></code> <code><a href="get_ideal_mat.html">add_ideal_mat()</a></code>

        </dt>
        <dd>Get ideal bin matrix</dd>
      </dl><dl><dt>

          <code><a href="get_snp_bidx.html">get_snp_bidx()</a></code>

        </dt>
        <dd>Get SNP bin idx</dd>
      </dl><dl><dt>

          <code><a href="get_snp_counts.html">get_snp_counts()</a></code>

        </dt>
        <dd>get_snp_counts</dd>
      </dl><dl><dt>

          <code><a href="get_tiled_bins.html">get_tiled_bins()</a></code>

        </dt>
        <dd>Get tiled bins</dd>
      </dl><dl><dt>

          <code><a href="grab_hmm_res.html">grab_hmm_res()</a></code>

        </dt>
        <dd>Grab HMM results from returned list</dd>
      </dl><dl><dt>

          <code><a href="hmmcopy_singlecell.html">hmmcopy_singlecell()</a></code>

        </dt>
        <dd>Single Cell HMMcopy</dd>
      </dl><dl><dt>

          <code><a href="identify_normal.html">identify_normal()</a></code>

        </dt>
        <dd>Identify normal cells from scATAC data</dd>
      </dl><dl><dt>

          <code><a href="integrate_segments.html">integrate_segments()</a></code>

        </dt>
        <dd>Integrate segments between assays</dd>
      </dl><dl><dt>

          <code><a href="is_ideal_bin.html">is_ideal_bin()</a></code>

        </dt>
        <dd>Flag ideal bins</dd>
      </dl><dl><dt>

          <code><a href="leiden_wrapper.html">leiden_wrapper()</a></code>

        </dt>
        <dd>Wrapper for the Leiden Algorithm</dd>
      </dl><dl><dt>

          <code><a href="length_normalize.html">length_normalize()</a></code>

        </dt>
        <dd>Length normalize counts</dd>
      </dl><dl><dt>

          <code><a href="load_atac_bins.html">load_atac_bins()</a></code>

        </dt>
        <dd>Load atac binned depth data</dd>
      </dl><dl><dt>

          <code><a href="logNorm.html">logNorm()</a></code>

        </dt>
        <dd>logNorm()</dd>
      </dl><dl><dt>

          <code><a href="logr_col_fun.html">logr_col_fun()</a></code>

        </dt>
        <dd>Log Ratio Colors</dd>
      </dl><dl><dt>

          <code><a href="merge_segments.html">merge_segments()</a></code>

        </dt>
        <dd>Merge segment levels</dd>
      </dl><dl><dt>

          <code><a href="overlap_genes.html">overlap_genes()</a></code>

        </dt>
        <dd>Overlap genes with bins</dd>
      </dl><dl><dt>

          <code><a href="params_sc_hmm.html">params_sc_hmm()</a></code>

        </dt>
        <dd>Single Cell HMMcopy parameters</dd>
      </dl><dl><dt>

          <code><a href="perform_gc_cor.html">perform_gc_cor()</a></code>

        </dt>
        <dd>Perform GC Correction</dd>
      </dl><dl><dt>

          <code><a href="phase_snps.html">phase_snps()</a></code>

        </dt>
        <dd>EM phasing of SNPs</dd>
      </dl><dl><dt>

          <code><a href="plot_cell_cna.html">plot_cell_cna()</a></code>

        </dt>
        <dd>Plot Cell Copy Number</dd>
      </dl><dl><dt>

          <code><a href="plot_cell_multi.html">plot_cell_multi()</a></code>

        </dt>
        <dd>Plot multiple cell assays together</dd>
      </dl><dl><dt>

          <code><a href="plot_cell_psuedobulk_cna.html">plot_cell_psuedobulk_cna()</a></code>

        </dt>
        <dd>Plot Psuedobulk cell CNA profiles</dd>
      </dl><dl><dt>

          <code><a href="plot_clone_comp.html">plot_clone_comp()</a></code>

        </dt>
        <dd>Clone Copy Number Comparison Plot</dd>
      </dl><dl><dt>

          <code><a href="plot_gene_cna.html">plot_gene_cna()</a></code>

        </dt>
        <dd>Plot Gene CNA</dd>
      </dl><dl><dt>

          <code><a href="plot_segs.html">plot_segs()</a></code>

        </dt>
        <dd>Plot data with segments</dd>
      </dl><dl><dt>

          <code><a href="pseudo_groups.html">pseudo_groups()</a></code>

        </dt>
        <dd>Pseudobulk groups</dd>
      </dl><dl><dt>

          <code><a href="pseudo_join.html">pseudo_join()</a></code>

        </dt>
        <dd>Join SNP and scATAC depth</dd>
      </dl><dl><dt>

          <code><a href="pseudobulk_sce.html">pseudobulk_sce()</a></code>

        </dt>
        <dd>Pseudobulk cell CNA profiles</dd>
      </dl><dl><dt>

          <code><a href="read_vartrix.html">read_vartrix()</a></code>

        </dt>
        <dd>Read Vatrix Data</dd>
      </dl><dl><dt>

          <code><a href="rebin_sce.html">rebin_sce()</a></code>

        </dt>
        <dd>Rebin <code>SingleCellExperiment</code> object</dd>
      </dl><dl><dt>

          <code><a href="run_sc_hmmcopy.html">run_sc_hmmcopy()</a></code>

        </dt>
        <dd>Single Cell HMMcopy</dd>
      </dl><dl><dt>

          <code><a href="run_scatools.html">run_scatools()</a></code>

        </dt>
        <dd>Run scATAC Copy Number Analysis</dd>
      </dl><dl><dt>

          <code><a href="scale_sub.html">scale_sub()</a></code>

        </dt>
        <dd>Scale and subset SCE assays</dd>
      </dl><dl><dt>

          <code><a href="segment_cnv.html">segment_cnv()</a></code>

        </dt>
        <dd>scATAC CBS CNV segmentation</dd>
      </dl><dl><dt>

          <code><a href="smooth_counts.html">smooth_counts()</a></code>

        </dt>
        <dd>Smooth outlier copy number counts</dd>
      </dl><dl><dt>

          <code><a href="summarise_chr_arm.html">summarise_chr_arm()</a></code>

        </dt>
        <dd>Summarise copy number signal over chromosome arms</dd>
      </dl><dl><dt>

          <code><a href="test_sce.html">test_sce</a></code>

        </dt>
        <dd>Processed scATAC dataset</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Matthew Zatzman.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

